### PR TITLE
Update (ci): github action's artifact upgrade v4

### PR DESCRIPTION
**Pull request by alleeclark**
## What does this PR do?

- Updates GitHub actions's artifacts from v3 to v4
- Enable overwrite option on upload since dist packages has a placeholder uploaded on init followed by build packages. This version of upload restricts that -- see below in breaking changes
- Set merge-multiple to true to so downloaded artifacts will get written into the same directory
- Add https://developer.habana.ai to the ignore list for link checker. The developer blog redirects to intel's developer website. It receives client errors on the redirect and it reports as broken in link checker. 

## Why
We received the email below stating that we need to upgrade by January.

image

[Here is the recommended upgrade guide that I followed. 
](https://github.blog/news-insights/product-news/get-started-with-v4-of-github-actions-artifacts/#compatibility)

### Does this introduce breaking change?

-  > In v4, Artifacts are immutable (unless deleted). So you must change each of the uploaded Artifacts to have a different name and filter the downloads by name to achieve the same effect:

- - Since we are only uploading one artifact, it is fine. If we introduce a list, we will need to modify the name.

- According to Github after Jan 9th if we have not upgraded there will be issues running pipelines.

## Test Plan
- CI test
- Downloaded artifacts from a previous run to ensure the structure was the same
-- [dist-packages-870cec2e4fc4ca4f9ea49a94862b7412bd914431.zip](https://github.com/user-attachments/files/18156310/dist-packages-870cec2e4fc4ca4f9ea49a94862b7412bd914431.zip)
-- [dist-packages-e9100e548a5f90fb4e90a189b6f6c80dae4a37e7.zip](https://github.com/user-attachments/files/18156311/dist-packages-e9100e548a5f90fb4e90a189b6f6c80dae4a37e7.zip)

<details>
  <summary><b>Before submitting</b></summary>
- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20500.org.readthedocs.build/en/20500/

cc @borda

### CODE REVIEW

File: .github/workflows/_build-packages.yml [hunk 1]
```diff
@@ -63,8 +64,9 @@ jobs:
           mkdir pypi/${{ matrix.pkg-name }}
           cp dist/* pypi/${{ matrix.pkg-name }}/
 
-      - uses: actions/upload-artifact@v3
+      - uses: actions/upload-artifact@v4
         with:
           name: ${{ inputs.artifact-name }}
           path: pypi
           include-hidden-files: true
+          overwrite: true
```
Comment (alleeclark) [inline .github/workflows/_build-packages.yml:72]: Artifacts are uploaded in the init job (line 28) as a placeholder, therefore we need to overwrite them. See https://github.com/Lightning-AI/pytorch-lightning/actions/runs/12359005253/job/34490799219#step:8:22

File: docs/source-pytorch/conf.py [hunk 2]
```diff
@@ -643,6 +643,7 @@ def package_list_from_file(file):
     r"starter/installation.html$",
     r"^../common/trainer.html#trainer-flags$",
     "https://deepgenerativemodels.github.io/assets/slides/cs236_lecture11.pdf",
+    "https://developer.habana.ai", # returns 403 error but redirects to intel.com documentation
```
Comment (alleeclark) [inline docs/source-pytorch/conf.py:646]: [some habana.ai  sub domains route to intel](https://habana.ai/press/intel-acquires-artificial-intelligence-chipmaker-habana-labs/). For some reason this just started causing 403 errors.

Comment (alleeclark) [reply docs/source-pytorch/conf.py:646]: https://developer.habana.ai redirects to https://www.intel.com/content/www/us/en/developer/platform/gaudi/overview.html

**Review by Borda — CHANGES_REQUESTED** [review]
The direction is good, it just needs some more work with adding patterns and making unique uploads... 
pls, have look what I gave done earlier this week for shared utils project: https://github.com/Lightning-AI/utilities/pull/344

File: .github/workflows/ci-pkg-install.yml [hunk 3]
```diff
@@ -50,10 +50,11 @@ jobs:
       - uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.python-version }}
-      - uses: actions/download-artifact@v3
+      - uses: actions/download-artifact@v4
         with:
           name: dist-packages-${{ github.sha }}
           path: dist
+          merge-multiple: true
```
Comment (Borda) [inline .github/workflows/ci-pkg-install.yml:57]:
```suggestion
```

File: .github/workflows/docs-build.yml [hunk 4]
```diff
@@ -158,10 +158,11 @@ jobs:
       # use input if dispatch or git tag
       VERSION: ${{ inputs.version || github.ref_name }}
     steps:
-      - uses: actions/download-artifact@v3
+      - uses: actions/download-artifact@v4
         with:
           name: docs-${{ matrix.pkg-name }}-${{ github.sha }}
           path: docs/build/html/
+          merge-multiple: true
```
Comment (Borda) [inline .github/workflows/docs-build.yml:165]:
```suggestion
```

File: .github/workflows/release-pkg.yml [hunk 5]
```diff
@@ -38,10 +38,11 @@ jobs:
     if: github.event_name == 'release'
     steps:
       - uses: actions/checkout@v4
-      - uses: actions/download-artifact@v3
+      - uses: actions/download-artifact@v4
         with:
           name: dist-packages-${{ github.sha }}
           path: dist
+          merge-multiple: true
```
Comment (Borda) [inline .github/workflows/release-pkg.yml:45]:
```suggestion
```

File: .github/workflows/release-pkg.yml [hunk 6]
```diff
@@ -140,10 +141,11 @@ jobs:
         name: ["FABRIC", "PYTORCH", "LIGHTNING"]
     steps:
       - uses: actions/checkout@v4 # needed for local action below
-      - uses: actions/download-artifact@v3
+      - uses: actions/download-artifact@v4
         with:
           name: dist-packages-${{ github.sha }}
           path: dist
+          merge-multiple: true
```
Comment (Borda) [inline .github/workflows/release-pkg.yml:148]:
```suggestion
```

File: .github/workflows/release-pkg.yml [hunk 7]
```diff
@@ -165,10 +167,11 @@ jobs:
...
```
Comment (Borda) [inline .github/workflows/release-pkg.yml:174]:
```suggestion
```

**Review by Borda — APPROVED** [review]

**Comment by alleeclark** [comment]
> The direction is good, it just needs some more work with adding patterns and making unique uploads... pls, have look what I gave done earlier this week for shared utils project: [Lightning-AI/utilities#344](https://github.com/Lightning-AI/utilities/pull/344)

Thanks. I reviewed the changes.

**Pull request merged**